### PR TITLE
Add Result type

### LINKED ISSUE #80: Add a Result type

### PR DESCRIPTION
**Pull request by thinker227**
Add a minimal `Result<T, TError>` type. Closes #80.

### CODE REVIEW

File: src/Draco.Compiler.Internal/Utilities/Result.cs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+using System.Diagnostics.CodeAnalysis;
+
+namespace Draco.Compiler.Internal.Utilities;
+
+/// <summary>
+/// A union type containing either the result of an operation or an error.
+/// </summary>
+/// <typeparam name="T">The type of the inner value.</typeparam>
+/// <typeparam name="TError">The type of the inner error.</typeparam>
+public readonly struct Result<T, TError>
```
Comment (LPeter1997) [inline src/Draco.Compiler.Internal/Utilities/Result.cs:11]: Maybe TOk and TError? T and TError together sound very confusing.

File: src/Draco.Compiler.Internal/Utilities/Result.cs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    private readonly T value;
+    private readonly TError error;
+    private readonly bool isSuccess;
+
+    /// <summary>
+    /// The value of the result. Throws an exception if the result is not a value.
+    /// </summary>
+    /// <exception cref="InvalidOperationException"></exception>
+    public T? Value =>
+        this.isSuccess
```
Comment (LPeter1997) [inline src/Draco.Compiler.Internal/Utilities/Result.cs:22]: We usually keep the condition in the same line for ternary, less overall indentation

File: src/Draco.Compiler.Internal/Utilities/Result.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    /// <summary>
+    /// Whether the result is a successful value or not.
+    /// </summary>
+    [MemberNotNullWhen(true, nameof(Value))]
+    [MemberNotNullWhen(false, nameof(Error))]
+    public bool IsSuccess =>
+        this.isSuccess;
+
+
```
Comment (LPeter1997) [inline src/Draco.Compiler.Internal/Utilities/Result.cs:43]: Lot of stray newline

File: src/Draco.Compiler.Internal/Utilities/Result.cs [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    /// <exception cref="InvalidOperationException"></exception>
+    public TError? Error =>
+        this.isSuccess
+            ? throw new InvalidOperationException("Result does not contain an error.")
+            : this.error;
+
+    /// <summary>
+    /// Whether the result is a successful value or not.
+    /// </summary>
+    [MemberNotNullWhen(true, nameof(Value))]
```
Comment (LPeter1997) [inline src/Draco.Compiler.Internal/Utilities/Result.cs:38]: That's a crazy good trick, love it

File: src/Draco.Compiler.Internal/Utilities/Result.cs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+/// <summary>
+/// A union type containing either the result of an operation or an error.
+/// </summary>
+/// <typeparam name="TOk">The type of the inner ok value.</typeparam>
+/// <typeparam name="TError">The type of the inner error.</typeparam>
+public readonly struct Result<TOk, TError>
+{
+    private readonly TOk value;
+    private readonly TError error;
+    private readonly bool isSuccess;
```
Comment (LPeter1997) [inline src/Draco.Compiler.Internal/Utilities/Result.cs:15]: Please tidy up the names related to the ok value then, like isOk

**Comment by LPeter1997** [comment]
Some safe retrieval API would also be nice, like `OkOr(default)` and `ErrorOr(default)`. They could also accept value factory funcs.

File: src/Draco.Compiler.Internal/Utilities/Result.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        : new(this.error);
+
+    /// <summary>
+    /// Binds the error value to a new result with another error type.
+    /// </summary>
+    /// <typeparam name="UError">The type of the new error value.</typeparam>
+    /// <param name="transform">A transformation function mapping the error value to a new result.</param>
+    /// <returns>A new result of either the returned result or the previous ok value.</returns>
+    public Result<TOk, UError> BindError<UError>(Func<TError, Result<TOk, UError>> transform) =>
+        this.isOk
```
Comment (LPeter1997) [inline src/Draco.Compiler.Internal/Utilities/Result.cs:100]: Missed ternary, or is the line too long?

File: src/Draco.Compiler.Internal/Utilities/Result.cs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    public TError? Error => this.isOk
+        ? throw new InvalidOperationException("Result is not an error.")
+        : this.error;
+
+    /// <summary>
+    /// Whether the result is an ok value or not.
+    /// </summary>
+    [MemberNotNullWhen(true, nameof(Ok))]
+    [MemberNotNullWhen(false, nameof(Error))]
+    public bool IsOk =>
```
Comment (LPeter1997) [inline src/Draco.Compiler.Internal/Utilities/Result.cs:38]: Why not make it an auto-prop with a get-only accessor?

File: src/Draco.Compiler.Internal/Utilities/Result.cs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+/// <summary>
+/// A union type containing either the result of an operation or an error.
+/// </summary>
+/// <typeparam name="TOk">The type of the inner ok value.</typeparam>
+/// <typeparam name="TError">The type of the inner error.</typeparam>
+public readonly struct Result<TOk, TError>
+{
+    private readonly TOk ok;
+    private readonly TError error;
+    private readonly bool isOk;
```
Comment (LPeter1997) [inline src/Draco.Compiler.Internal/Utilities/Result.cs:15]: Prefer fields below props, above ctor.

**Review by LPeter1997 — COMMENTED** [review]
Nice, last couple of nit.

File: src/Draco.Compiler.Internal/Utilities/Result.cs (hunk 6 above)
Comment (thinker227) [reply src/Draco.Compiler.Internal/Utilities/Result.cs:100]: missed

**Review by LPeter1997 — APPROVED** [review]

**Pull request merged**